### PR TITLE
Add wmstruct

### PR DESCRIPTION
**Pull request by ukos-git**
# Add WM Structure recognition.
fixes https://github.com/byte-physics/igor-code-browser/issues/10 

This will add support for the following parameters:
* `WMWinHookStruct`
* `WMBackgroundStruct`
* *any* other `WM.*Struct`

# Tested using

```igorpro
Structure WMtestStruct
EndStructure

Function MyNormalFunction(s)
	STRUCT WMtestStruct &s
End

Function MyWindowHook(s)
	STRUCT WMWinHookStruct &s

	Variable hookResult = 0

	switch(s.eventCode)
		case 0:				// Activate
			// Handle activate
			break

		case 1:				// Deactivate
			// Handle deactivate
			break

		// And so on . . .
	endswitch

	return hookResult		// 0 if nothing done, else 1
End

Function TestTask(s)		// This is the function that will be called periodically
	STRUCT WMBackgroundStruct &s
	
	Printf "Task %s called, ticks=%d\r", s.name, s.curRunTicks
	return 0	// Continue background task
End

Function StartTestTask()
	Variable numTicks = 2 * 60		// Run every two seconds (120 ticks)
	CtrlNamedBackground Test, period=numTicks, proc=TestTask
	CtrlNamedBackground Test, start
End

```

produces the followng decls:

```
\k(0,0,0)\K(0,0,0)\W522	WMtestStruct
\k(0,0,0)\K(0,0,0)\W529	[WMtestStruct] MyNormalFunction(struct&) -> var
\k(0,0,0)\K(0,0,0)\W529	[WMWinHookStruct] MyWindowHook(struct&) -> var
\k(0,0,0)\K(0,0,0)\W529	[WMBackgroundStruct] TestTask(struct&) -> var
\k(0,0,0)\K(0,0,0)\W529	StartTestTask() -> var
```

### CODE REVIEW

File: procedures/CodeBrowser.ipf [hunk 1]
```diff
@@ -211,6 +211,42 @@ Function/S interpretParameters(funcInfo)
...
+	string &WMDotStarStruct
+
+	string funcName, module, procedure, declaration, re
+	string handledWMStructs = "WinHook|Background|[^\s,]+" // WM + handledWMStructs + Struct
+
+	WMDotStarStruct = ""
+	if(NumberByKey("PARAM_0_TYPE", funcInfo) & 0x200 | 0x1000) // struct | pass-by-reference
+		funcName = StringByKey("NAME", funcInfo)
+		module = StringByKey("INDEPENDENTMODULE", funcInfo)
+		//module = strlen(module) == 0 ? StringByKey("MODULE", funcInfo) : module
```
Comment (ukos-git) [inline procedures/CodeBrowser.ipf]: delete?

Comment (t-b) [reply procedures/CodeBrowser.ipf]: @ukos-git Yes!

Comment (ukos-git) [reply procedures/CodeBrowser.ipf]: dropped in 
a918417

**Comment by t-b** [comment]
@ukos-git Works nicely. I'll look at the code later.

Just a quick comment about the design of the newly added structure name.

I think I would prefer something like

```
\k(0,0,0)\K(0,0,0)\W522	WMtestStruct
\k(0,0,0)\K(0,0,0)\W529	MyNormalFunction(WMtestStruct &) -> var
\k(0,0,0)\K(0,0,0)\W529	MyWindowHook(WMWinHookStruct &) -> var
\k(0,0,0)\K(0,0,0)\W529	TestTask(WMBackgroundStruct &) -> var
\k(0,0,0)\K(0,0,0)\W529	StartTestTask() -> var
```

as that is conciser. @MichaelHuth Thougts/Opinions?

How difficult is it to extend it to all functions taking a single structure argument?

**Comment by ukos-git** [comment]
That looks better. Not so easy because it would need a hack into the Parse-Arguments but it should be possible.

**Comment by t-b** [comment]
@ukos-git Please assing me/request a review once I should have a look again.

**Comment by ukos-git** [comment]
working on it right now.

**Comment by ukos-git** [comment]
> How difficult is it to extend it to all functions taking a single structure argument?

The matching pattern for the first line after the function definition looks like this right now:

```igorpro
string handledWMStructs = "WinHook|Background|[^\s,]+" // WM + handledWMStructs + Struct
re = "(?i)\s*struct\s+(WM(?:" + handledWMStructs + ")Struct)\s+.*"
```
do you mean, changing it to something like this:
```diff
- re = "(?i)\s*struct\s+(WM(?:" + handledWMStructs + ")Struct)\s+.*"
+ re = "(?i)^\s*struct\s+(\w+)\s+.*$"

**Comment by ukos-git** [comment]
/assign @t-b 
(cannot assign)

**Comment by MichaelHuth** [comment]
> as that is conciser. @MichaelHuth Thougts/Opinions?

I think it is good if it matches the current layout.

Though, I would prefer an overall 'table style' layout more with defined vertical and horizontal lines.

**Comment by t-b** [comment]
/assign @t-b 

(I've changed some permissions now we all are allowed to write the repos.)

File: procedures/CodeBrowser.ipf [hunk 2]
```diff
@@ -879,18 +879,17 @@ Function/S nicifyProcedureList(list)
...
 
-	if(isProcGlobal(module))
-		debugPrint(module + " is in ProcGlobal")
-		strProcedure = ProcedureText("", 0, procedureWithoutModule)
-		return strProcedure
-	else
-		debugPrint(procedureWithoutModule + " is in " + module)
-		return ProcedureText("", 0, procedureWithoutModule + " [" + module + "]")
+	if(!isProcGlobal(module))
+		procedure = procedure + "[" + module + "]"
```
Comment (t-b) [inline procedures/CodeBrowser.ipf]: the original version has a space before `[`.

File: procedures/CodeBrowser.ipf [hunk 3]
```diff
@@ -211,6 +217,38 @@ Function/S interpretParameters(funcInfo)
...
+	string funcInfo
+	string &WMDotStarStruct
+
+	string declaration, re, typeName
+	string handledWMStructs = "WinHook|Background|[^\s,]+" // WM + handledWMStructs + Struct
+
+	if(NumberByKey("PARAM_0_TYPE", funcInfo) & 0x200 | 0x1000) // struct | pass-by-reference
+		declaration = getFunctionLine(1, funcInfo)
+		re = "(?i)\s*struct\s+(WM(?:" + handledWMStructs + ")Struct)\s+.*"
+		SplitString/E=(re) declaration, typeName
```
Comment (t-b) [inline procedures/CodeBrowser.ipf]: How about making it viable for all structures? You already proposed a regexp which I used

```
$ git diff .
diff --git a/procedures/CodeBrowser.ipf b/procedures/CodeBrowser.ipf
index 98d78a8..ecf9105 100644
--- a/procedures/CodeBrowser.ipf
+++ b/procedures/CodeBrowser.ipf
@@ -234,11 +234,10 @@ Function FunctionWithWMstruct(funcInfo, WMDotStarStruct)
        string &WMDotStarStruct

        string declaration, re, typeName
-       string handledWMStructs = "WinHook|Background|[^\s,]+" // WM + handledWMStructs + Struct

-       if(NumberByKey("PARAM_0_TYPE", funcInfo) & 0x200 | 0x1000) // struct | pass-by-reference
+       if(NumberByKey("PARAM_0_TYPE", funcInfo) & (0x200 | 0x1000)) // struct | pass-by-reference
                declaration = getFunctionLine(1, funcInfo)
-               re = "(?i)\s*struct\s+(WM(?:" + handledWMStructs + ")Struct)\s+.*"
+               re = "(?i)^\s*struct\s+(\w+)\s+"
                SplitString/E=(re) declaration, typeName
                if(V_flag == 1)
                        WMDotStarStruct = typeName

```

and that works for

```
Structure MyStruct
EndStructure

Function ConsumeStruct(s)
        STRUCT MyStruct &s

        return 0
End
```

already.

In my patch I also added some parantheses, the precedence rules for `&` and `|` are a bit difficult to know by heart.

File: procedures/CodeBrowser.ipf [hunk 4]
```diff
@@ -117,7 +119,11 @@ Function/S interpretParamType(ptype, paramOrReturn)
 	elseif(ptype & 0x100)
 		typeStr += "dfref"
 	elseif(ptype & 0x200)
-		typeStr += "struct"
+		typeName = "struct"
+		if(FunctionWithWMstruct(funcInfo, typeName))
+			debugPrint("Found WaveMetrics Struct " + typeName)
+		endif
+		typeStr += typeName
```
Comment (t-b) [inline procedures/CodeBrowser.ipf]: That is rather difficult to grasp as you set `typeName` and then overwrite it.

How about

```
		if(FunctionWithWMstruct(funcInfo, typeName))
                        typeStr += typeName
			debugPrint("Found WaveMetrics Struct " + typeName)
                 else
                         typeStr += "struct"
		endif
```

File: procedures/CodeBrowser.ipf [hunk 5]
```diff
@@ -878,19 +916,52 @@ Function/S nicifyProcedureList(list)
...
+	module = StringByKey("INDEPENDENTMODULE", funcInfo)
+	procedure = StringByKey("PROCWIN", funcInfo)
+
+	if(!isProcGlobal(module))
+		debugPrint(procedure + " is not in ProcGlobal")
+		procedure = procedure + "[" + module + "]"
+	endif
+
+	linesOfContext = lineNo < 0 ? lineNo : 0
+	context = ProcedureText(funcName, linesOfContext, procedure)
```
Comment (t-b) [inline procedures/CodeBrowser.ipf]: You should reuse `getProcedureText` here which then needs an optional linesOfContext parameter. In that way you don't need to do the module handling here as well.

File: procedures/CodeBrowser.ipf (hunk 2 above)
Comment (ukos-git) [reply procedures/CodeBrowser.ipf]: that was definitely a mistake.

File: procedures/CodeBrowser.ipf (hunk 3 above)
Comment (ukos-git) [reply procedures/CodeBrowser.ipf]: it works without matching the last `.*` so I kept it like you proposed.

**Pull request merged**